### DISCUSSION TITLE
Highway mode?

### DISCUSSION THREAD

**Discussion by endolith** [streetcomplete/StreetComplete#4999] [Ideas]
Was trying to use this while riding in the car but by the time it loads a block of quests, I'm already past them. It would be good if it saw that I was moving fast and only loaded quests within a certain distance of the road I'm on, and far ahead of where I actually am

**Comment by Helium314** [comment]
That will not help, most of the time is spent on downloading and persisting data. Scanning that data for quests is quite fast, so you will barely notice the difference beween creating all quest types, or only a few types.

You can try StreetComplete pre-32.0, where quests were generated using overpass. Or manually downloading the area in advance if you know where the car will be going.

**Reply by endolith** [reply]
> That will not help, most of the time is spent on downloading and persisting data.

Yes, that's why I posted this.  It spends too much time on downloading data, so by the time it downloads an area, I'm already past it and can't complete any quests in it.  I'm saying it would be more useful if it only downloaded smaller areas near where I will be in the future, so by the time I get to them, they are downloaded and I can do quests in them.

**Reply by mnalis** [reply]
@endolith if you want to play with that, it is relatively easy even with no programming background. Click on edit (`pen` icon) on this file, and edit constants to your liking:https://github.com/streetcomplete/StreetComplete/blob/42771a9508865d3c49da3c7e2485c94d9814c34d/app/src/main/java/de/westnordost/streetcomplete/data/download/strategy/MobileDataAutoDownloadStrategy.kt#L12-L13
 
Github will ask you to create a project fork and then allow you to edit the file in your fork. 
After you've edited it, in your fork click on `Actions` tab, and then `Build debug apk` and `Run workflow`, where you choose the branch in which you edited constants to your liking. Wait a dozen minutes, and GitHub will build you an `.apk` file which you can download on your mobile phone and try out how it works. 

If you do try it out, let us know how well it worked for you. Did different value solved your problem completely, i.e. was it fast enough with smaller area? Or was it still too slow even with significantly reduced area? Or was it fast enough, but area was too small so it didn't catch quests you wanted?

**Reply by Helium314** [reply]
My very simple workaround for this is downloading the area where I will be manually, instead of hoping that SC will predict my movement correctly.

**Reply by endolith** [reply]
> @endolith if you want to play with that, it is relatively easy even with no programming background. Click on edit (`pen` icon) on this file, and edit constants to your liking:
> 
> https://github.com/streetcomplete/StreetComplete/blob/42771a9508865d3c49da3c7e2485c94d9814c34d/app/src/main/java/de/westnordost/streetcomplete/data/download/strategy/MobileDataAutoDownloadStrategy.kt#L12-L13

What do those do?  They just increase the search radius?  I want to _reduce_ the search radius so it downloads faster, and in areas I will be in the future, not where I am now.

**Reply by mnalis** [reply]
> What do those do? They just increase the search radius? 

They **define** the search radius. 
So if you **increase** them, then the search radius will **increase**, if you **decrease** them, then it will **decrease**.

- `maxDownloadAreaInKm2` is the limit of **maximum** area that can be downloaded in one request. So if you reduce it from `10.0` to `2.0`, it will _never_ download more than 2 km<sup>2</sup> (whereas before that change in default SC it could download _up to_ 10 km<sup>2</sup>)
- note that the above number is **maximum** size that SC will download. It may decide to download **less**, depending on the value of `desiredScoredMapDataCountInVicinity`

> I want to reduce the search radius so it downloads faster, and in areas I will be in the future, not where I am now.

SC (nor anybody else; if we exclude certain religious beliefs) does not **_know_** where you will be in the **future**. They only know where you are **now** (and where you've been in the past).

So if you for example reduce both numbers by a factor of 10 (i.e. making them `1.0` and `750` respectively), it will download about 10 times less data, which should be _faster_ (it would **not** be 10 times faster, though, as there are some fixed and non-linear delays depending on your network and CPU performance). 

You can try it and check how well it works for you, and then try to change them again to be even better for you (either increasing them a little, or reducing them even further). 

Let us know how how that worked, if you decide to try it out.

---

Alternative, as  suggested above by Helium314, is that as you yourself may have a pretty good (but still not perfect) idea where you might be in some seconds in the future, is  to disable auto download and just position the map manually and download the data manually. Or even pre-download quest data before you go on the trip.

**Reply by endolith** [reply]
> does not know where you will be in the future. They only know where you are now (and where you've been in the past).

Yes, but it can extrapolate that path to guess where I will be in the future, which will be useful 90% of the time, which is much better than loading data for where I am right now, which is useful 10% of the time, because I likely won't be there anymore by the time the data loads (unless we hit a traffic jam).

**Reply by mnalis** [reply]
> Yes, but it can extrapolate that path to guess where I will be in the future, which will be useful 90% of the time, 

If 90% of roads in your area happen to be straight lines, yes.

> which is much better than loading data for where I am right now, which is useful 10% of the time, because I likely won't be there anymore by the time the data loads (unless we hit a traffic jam).

Sure. But I've tried to give you help on simple thing that you can easily try yourself, so you provide feedback which might help move this forward. 

The more complex approach you mention here was talked about in the issue I linked, and main developed stated they [won't be implementing it themselves, but that other programmers are free to try to implement it](https://github.com/streetcomplete/StreetComplete/issues/3052#issuecomment-879160521) and provide feedback. But nobody stepped up to try it (so far). Which is somewhat understandable of course - even without programming the algorithm, it needs a lot of prep work measuring and testing, and the end result might still not be much better than current situation. The idea is that non-programmers could help with that prep work and testing.

**Reply by mnalis** [reply]
> which is much better than loading data for where I am right now

Regarding that: Not necessarily. Because that StreetComplete is intended to be **used primarily by pedestrians**, so it is reasonable assumption to make in such case as one won't get very far away on foot (see PR https://github.com/streetcomplete/StreetComplete/pull/4188 and linked issues)

**Reply by endolith** [reply]
> If 90% of roads in your area happen to be straight lines, yes.

Well I should hope they are, at the high speeds where this feature is needed, or else we would all crash all the time. 😁

> Sure. But I've tried to give you help on simple thing that you can easily try yourself, so you provide feedback which might help move this forward. 

It's only been a few days. I haven't even been a passenger in a car in weeks... 

Anyway, I don't understand how changing the search radius would help, can you explain why you think it would? Do you think decreasing it would help because of faster download, or increasing it would help because it would get quests further down the road?

**Reply by mnalis** [reply]
> how changing the search radius would help, can you explain why you think it would?

Well, if we knew the answer, nobody would need to test it, and we could just change the values so they work for everyone :smiley: 

Maybe the **smaller** values will make it be so fast that it will grab enough data in time while you're still there. But maybe even with very small values it is still too slow; then perhaps with **bigger** values it will be somewhat slower, but grab enough of surrounding area that even with that slowness you would still be in downloaded area by the time it finished downloading. And maybe neither would help. 

It's hard to say what might help and what might make situation worse without _actually_ trying it out (there are several non-linear effects acting in combination, some make it faster and some slower, and changing radius changes how both behave). So one interested in helping should build several different versions (little slower, little faster, much slower, much faster - e.g. using factor "2" for little, and "10" for much) and try them out and make videos and/or take timings of how they perform in various scenarios (e.g. urban residential road vs. motorway) and compare that to current defaults. 

That is relatively easy to do for non-programmers, but is quite time consuming (if you just do the tests one build per ride, it can be 5*2=10 different rides, but if you optimize by pre-building all the different versions, you can do all the test in one or two bigger rides; uninstalling one version and installing another during the ride)...

That is why someone willing to invest some time to try few different combinations and test them and report results is needed. 
Even if the result is that neither increasing nor decreasing the radius helps fix the issue completely (which is relatively high probability, but not a certainty) it might make a better defaults, and even if not, it is still valuable data as the timings of bigger/smaller values done should help if some android programmer can dedicate some time to implement more advanced algorithm, but does not have time to do all that prep work.

> It's only been a few days. I haven't even been a passenger in a car in weeks...

I apologize, I didn't mean to sound harsh/rushing :heart:

**Reply by endolith** [reply]
Ok. I'm in a car now and watching it, and I'd say the default settings are good enough maybe half the time, while the rest of the time it loads too slowly. Taking some screenshots but I don't think I have an ability to record videos right now.

Traveling 60 mph

**Reply by endolith** [reply]
Also the "are you sure?" dialog is very annoying and I wish I could turn it off permanently.

**Reply by mnalis** [reply]
> Ok. I'm in a car now and watching it, and I'd say the default settings are good enough maybe half the time, while the rest of the time it loads too slowly

Useful information, thanks! Let us know how it handles when the range is reduced and extended when you get a chance to try it out.

> but I don't think I have an ability to record videos right now

I think at least Android 10+ should have `Screen recorder` in the drawer (when you swipe down from top of the screen), but you might have to click "Edit" button first to make it show by default. But vehicle speed + several screenshots (before loading, while loading and immediately after loading) help too (they might even show related info like what mobile network e.g. 2G/3G/4G/5G is in use) if video is not available. 

Geo locations of where it loads in time and where it doesn't (if not a privacy concern) would also be helpful, as amount of data to be downloaded is determined based on small "pilot/sample" download around your current location (to help estimate element density and thus tune size of bigger "real" download) 

> Also the "are you sure?" dialog is very annoying and I wish I could turn it off permanently.

you mean _"Are you sure you checked this on-site"_ ? There is checkbox _"Don't show again for this session"_ that should help for being less annoying. But it is generally useful as it stops mistakes (happened to me few times too, and I'm not a newbie), and does help with wrongly adding `source=survey` (there is also related PR https://github.com/streetcomplete/StreetComplete/pull/4947 which might help with that popping up more reasonably, go there and your :+1: to it if you'd like that)

**Comment by mnalis** [comment]
- related issue: https://github.com/streetcomplete/StreetComplete/issues/3052
